modules/character_net_controller/character_net_controller.h

```diff
@@ -0,0 +1,426 @@
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "scene/main/node.h"
+
+#include "core/math/transform.h"
+#include "core/node_path.h"
+#include "input_buffer.h"
+#include "net_utilities.h"
+#include <deque>
+#include <vector>
+
+#ifndef CHARACTER_NET_CONTROLLER_H
+#define CHARACTER_NET_CONTROLLER_H
+
+struct Controller;
+class Spatial;
+
+class CharacterNetController : public Node {
+	GDCLASS(CharacterNetController, Node);
+
+public:
+	enum InputCompressionLevel {
+		INPUT_COMPRESSION_LEVEL_0 = InputsBuffer::COMPRESSION_LEVEL_0,
+		INPUT_COMPRESSION_LEVEL_1 = InputsBuffer::COMPRESSION_LEVEL_1,
+		INPUT_COMPRESSION_LEVEL_2 = InputsBuffer::COMPRESSION_LEVEL_2,
+		INPUT_COMPRESSION_LEVEL_3 = InputsBuffer::COMPRESSION_LEVEL_3
+	};
+
+private:
+	/// The snapshot storage size is used to cap the amount of inputs collected by
+	/// the `Master`.
+	///
+	/// The server sends a message, to all the connected peers, notifing its
+	/// status at a fixed interval.
+	/// The peers, after receiving this update, removes all the old inputs until
+	/// that moment.
+	///
+	/// If the `input_storage_size` is too small, the clients will collect inputs
+	/// intermittently, but on the other side, a too large value may introduce
+	/// virtual delay.
+	///
+	/// With 60 iteration per seconds a good value is `300`, but is adviced to
+	/// perform some tests until you find a better suitable value for your needs.
+	int player_snapshot_storage_size;
+
+	/// Used to set the amount of traced frames to determine the connection healt trend.
+	///
+	/// This parameter depends a lot on the physics iteration per second, and
+	/// an optimal parameter, with 60 physics iteration per second, is 1200;
+	/// that is equivalent of the latest 20 seconds frames.
+	///
+	/// A smaller value will make the recovery mechanism too noisy and so useless,
+	/// on the other hand a too big value will make the recovery mechanism too
+	/// slow.
+	int network_traced_frames;
+
+	/// Amount of time an inputs is re-sent to each node.
```

Comment:
typo? I guess this is input instead of inputs

Reply:
Nope, but peer sounds better.